### PR TITLE
Refactor the SDK helpers, create MeterImpl

### PR DESCRIPTION
**Pull request by jmacd**
This reduces the boilerplate needed to implement the `Meter`. `api/metric/sdkhelpers.go` now defines a 4-method `MeterImpl` interface which supports general-purpose constructors instead of specific ones, i.e., `NewSynchronousInstrument` and `NewAsynchronousInstrument`. The two implementations (the Mock and the real SDK) are simplified. The global Meter implementation is also refactored to share as much of possible of this code, however it is written not to require the use of `MeterImpl`, as it is just a helper. (An SDK could decide to implement the constructors directly.)

This refactoring is a precursor to #514, since after we have the four-method `MeterImpl` a name-uniqueness wrapper will simply wrap around the `MeterImpl` and provide uniqueness checking.

Removes the `Unregister()` method for observers, since it was not in the spec. This can be restored in another PR if needed, but I would prefer we ask users to shut down the SDK instead of unregistering observers.

### CODE REVIEW

File: api/global/internal/meter.go [hunk 1]
```diff
@@ -244,20 +206,27 @@ func (inst *instImpl) setDelegate(d metric.Meter) {
 	atomic.StorePointer(&inst.delegate, unsafe.Pointer(implPtr))
 }
 
-func (inst *instImpl) Bind(labels metric.LabelSet) metric.BoundInstrumentImpl {
-	if implPtr := (*metric.InstrumentImpl)(atomic.LoadPointer(&inst.delegate)); implPtr != nil {
+func (inst *syncImpl) Interface() interface{} {
```
Comment (jmacd) [inline api/global/internal/meter.go]: This and the corresponding async version need testing.

**Comment by jmacd** [comment]
There are a few tests still failing, but I will update those. This is ready for review.

File: api/metric/sdkhelpers.go [hunk 2]
```diff
@@ -29,20 +55,35 @@ type LabelSetDelegate interface {
...
 
-// InstrumentImpl is the implementation-level interface Set/Add/Record
-// individual metrics without precomputed labels.
+// InstrumentImpl is a common interface for synchronous and
+// asynchronous instruments.
 type InstrumentImpl interface {
+	// Interface returns the underlying implementation of the
+	// instrument, which allows the implementation to gain access
+	// to its own representation especially from a `Measurement`.
+	Interface() interface{}
```
Comment (Aneurysm9) [inline api/metric/sdkhelpers.go]: This name seems off to me.  My first thought is that this should be `Implementation() interface{}`.  I know that it returns an `interface{}`, but it seems that the intent is to access the underlying implementation.

Comment (jmacd) [reply api/metric/sdkhelpers.go]: :+1:

Comment (jmacd) [reply api/metric/sdkhelpers.go]: Done.

**Comment by jmacd** [comment]
Reviewers: see how I intend to build off of this functionality in https://github.com/open-telemetry/opentelemetry-go/pull/566

File: api/metric/sdkhelpers.go [hunk 3]
```diff
@@ -29,20 +55,35 @@ type LabelSetDelegate interface {
...
+	// Implementation returns the underlying implementation of the
+	// instrument, which allows the implementation to gain access
+	// to its own representation especially from a `Measurement`.
+	Implementation() interface{}
+
+	// Descriptor returns a copy of the instrument's Descriptor.
+	Descriptor() Descriptor
+}
+
+// SynchronousImpl is the implementation-level interface Set/Add/Record
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Drop the `Set/` part - leftover from gauges.

File: api/metric/sdkhelpers.go [hunk 4]
```diff
@@ -29,20 +55,35 @@ type LabelSetDelegate interface {
...
 	// precomputed labels.
-	Bind(labels LabelSet) BoundInstrumentImpl
+	Bind(labels LabelSet) BoundSynchronousImpl
 
 	// RecordOne allows the SDK to observe a single metric event.
 	RecordOne(ctx context.Context, number core.Number, labels LabelSet)
 }
 
-// BoundInstrumentImpl is the implementation-level interface to Set/Add/Record
+// BoundSynchronousImpl is the implementation-level interface to Set/Add/Record
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Drop the `Set/` part - leftover from gauges.

File: api/metric/sdkhelpers.go [hunk 5]
```diff
@@ -51,42 +92,38 @@ type BoundInstrumentImpl interface {
...
-	return Float64Measure{commonMetric: common}, err
+// float64ObserverResult is an adapter for float64-valued asynchronous
+// callbacks.
+type float64ObserverResult struct {
+	observe func(core.Number, LabelSet)
 }
 
+var (
+	_ Meter                 = (*wrappedMeterImpl)(nil)
+	_ Int64ObserverResult   = (*int64ObserverResult)(nil)
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Does it need to be a pointer?

File: api/metric/sdkhelpers.go [hunk 6]
```diff
@@ -51,42 +92,38 @@ type BoundInstrumentImpl interface {
...
+	_ Float64ObserverResult = (*float64ObserverResult)(nil)
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Does it need to be a pointer?

File: api/metric/sdkhelpers.go [hunk 7]
```diff
@@ -95,3 +132,146 @@ func Configure(opts []Option) Config {
...
+
+func (m *wrappedMeterImpl) RegisterInt64Observer(name string, callback Int64ObserverCallback, opts ...Option) (Int64Observer, error) {
+	if callback == nil {
+		return NoopMeter{}.RegisterInt64Observer("", nil)
+	}
+	result := &int64ObserverResult{}
+	return WrapInt64ObserverInstrument(
+		m.newAsynchronous(name, ObserverKind, core.Int64NumberKind, opts,
+			func(observe func(core.Number, LabelSet)) {
+				result.observe = observe
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Could the result be constructed inside the callback? If it implements the result interface by value, then there should be no extra allocation, right?

File: api/metric/sdkhelpers.go [hunk 8]
```diff
@@ -95,3 +132,146 @@ func Configure(opts []Option) Config {
...
+
+func (m *wrappedMeterImpl) RegisterFloat64Observer(name string, callback Float64ObserverCallback, opts ...Option) (Float64Observer, error) {
+	if callback == nil {
+		return NoopMeter{}.RegisterFloat64Observer("", nil)
+	}
+	result := &float64ObserverResult{}
+	return WrapFloat64ObserverInstrument(
+		m.newAsynchronous(name, ObserverKind, core.Float64NumberKind, opts,
+			func(observe func(core.Number, LabelSet)) {
+				result.observe = observe
```
Comment (krnowak) [inline api/metric/sdkhelpers.go]: Could the result be constructed inside the callback? If it implements the result interface by value, then there should be no extra allocation, right?

File: internal/metric/mock.go [hunk 9]
```diff
@@ -54,101 +48,83 @@ type (
...
-	int64ObserverResult struct {
-		result observerResult
-	}
+	Asynchronous struct {
+		Instrument
 
-	float64ObserverResult struct {
-		result observerResult
-	}
+		Dead bool
```
Comment (krnowak) [inline internal/metric/mock.go]: You could drop it - it was used for observer unregistration, which you removed.

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go [hunk 10]
```diff
@@ -56,7 +56,7 @@ func TestGroupingStateless(t *testing.T) {
 	// Repeat for {counter,lastvalue}.{1,2}.
 	// Output lastvalue should have only the "G=H" and "G=" keys.
 	// Output counter should have only the "C=D" and "C=" keys.
-	require.EqualValues(t, map[string]int64{
+	require.EqualValues(t, map[string]float64{
```
Comment (krnowak) [inline sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: Why this change? I don't get it.

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go [hunk 11]
```diff
@@ -95,7 +95,7 @@ func TestGroupingStateful(t *testing.T) {
 	err := checkpointSet.ForEach(records1.AddTo)
 	require.NoError(t, err)
 
-	require.EqualValues(t, map[string]int64{
+	require.EqualValues(t, map[string]float64{
```
Comment (krnowak) [inline sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: Why this change? I don't get it.

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go [hunk 12]
```diff
@@ -138,7 +138,7 @@ func TestGroupingStateful(t *testing.T) {
 	err = checkpointSet.ForEach(records4.AddTo)
 	require.NoError(t, err)
 
-	require.EqualValues(t, map[string]int64{
+	require.EqualValues(t, map[string]float64{
```
Comment (krnowak) [inline sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: Why this change? I don't get it.

File: sdk/metric/batcher/test/test.go [hunk 13]
```diff
@@ -32,7 +34,7 @@ type (
 	Encoder struct{}
 
 	// Output collects distinct metric/label set outputs.
-	Output map[string]int64
+	Output map[string]float64
```
Comment (krnowak) [inline sdk/metric/batcher/test/test.go]: Why this change? I don't get it.

File: sdk/metric/batcher/test/test.go [hunk 14]
```diff
@@ -137,14 +139,16 @@ func CounterAgg(desc *export.Descriptor, v int64) export.Aggregator {
...
 	key := fmt.Sprint(rec.Descriptor().Name(), "/", labels.Encoded())
-	var value int64
-	switch t := rec.Aggregator().(type) {
-	case *sum.Aggregator:
-		sum, _ := t.Sum()
-		value = sum.AsInt64()
-	case *lastvalue.Aggregator:
-		lv, _, _ := t.LastValue()
-		value = lv.AsInt64()
+	var value float64
```
Comment (krnowak) [inline sdk/metric/batcher/test/test.go]: Was something wrong with `int64`?

File: sdk/metric/correct_test.go [hunk 15]
```diff
@@ -289,3 +296,45 @@ func TestDefaultLabelEncoder(t *testing.T) {
...
+func TestObserverCollection(t *testing.T) {
+	ctx := context.Background()
+	batcher := &correctnessBatcher{
+		t: t,
+	}
+	sdk := sdk.New(batcher, sdk.NewDefaultLabelEncoder())
+	meter := metric.WrapMeterImpl(sdk)
+
+	_ = Must(meter).RegisterFloat64Observer("float.observer", func(result metric.Float64ObserverResult) {
+		// TODO: The spec says the last-value wins in observer
```
Comment (krnowak) [inline sdk/metric/correct_test.go]: That could be my misunderstanding of what last-value aggregation within a labelset and measure aggregation across labelsets meant in the spec. Maybe file an issue about it and assign it to me?

File: sdk/metric/sdk.go [hunk 16]
```diff
@@ -117,36 +117,27 @@ type (
 		// labels is the LabelSet passed by the user.
 		labels *labels
 
-		// descriptor describes the metric instrument.
-		descriptor *export.Descriptor
+		//
```
Comment (krnowak) [inline sdk/metric/sdk.go]: Missing doc?

File: api/metric/api.go [hunk 17]
```diff
@@ -72,6 +74,74 @@ func (m Measurement) Number() core.Number {
...
+	// ObserverKind indicates an Observer instrument.
+	ObserverKind
+	// CounterKind indicates a Counter instrument.
+	CounterKind
+)
+
+// Descriptor contains all the settings that describe an instrument,
+// including its name, metric kind, number kind, and the configurable
+// options.
+type Descriptor struct {
```
Comment (krnowak) [inline api/metric/api.go]: So the prodigal son returns. ;)

File: api/metric/sdkhelpers.go (hunk 5 above)
Comment (jmacd) [reply api/metric/sdkhelpers.go]: I removed the use a pointer here. It doesn't need to be one way or another, but there is a small allocation as a result. (I believe a function pointer is greater than a single word, so the aforementioned single-word-object optimization would not have applied.)

File: api/metric/sdkhelpers.go (hunk 6 above)
Comment (jmacd) [reply api/metric/sdkhelpers.go]: (Done.)

File: api/metric/sdkhelpers.go (hunk 8 above)
Comment (jmacd) [reply api/metric/sdkhelpers.go]: There is an allocation, but I did it anyway. Let someone complain. :-)

File: internal/metric/mock.go (hunk 9 above)
Comment (jmacd) [reply internal/metric/mock.go]: Sorry about this. I'm happy to restore Unregister in another PR, but I don't think anyone else has mentioned it.

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go (hunk 10 above)
Comment (jmacd) [reply sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: I added a new test in `sdk/metric` because I had let a bug slip through and only the OTLP end-to-end test recently introduced actually tested that the real SDK would call the observers. The new test uses this helper.

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go (hunk 11 above)
Comment (jmacd) [reply sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: (None of the batcher tests were testing float instruments, I guess that's another reason to make the change.)

File: sdk/metric/batcher/defaultkeys/defaultkeys_test.go (hunk 12 above)
Comment (jmacd) [reply sdk/metric/batcher/defaultkeys/defaultkeys_test.go]: (Same ^^^)

File: sdk/metric/batcher/test/test.go (hunk 13 above)
Comment (jmacd) [reply sdk/metric/batcher/test/test.go]: (Same ^^^)

File: sdk/metric/batcher/test/test.go (hunk 14 above)
Comment (jmacd) [reply sdk/metric/batcher/test/test.go]: (Same ^^^)

File: sdk/metric/correct_test.go (hunk 15 above)
Comment (jmacd) [reply sdk/metric/correct_test.go]: I had forgotten which details are in the spec and which are in the OTEP that's not yet merged. This detail is not yet merged, so there's not really a problem. I filed a ticket #564.

File: api/metric/api.go (hunk 17 above)
Comment (jmacd) [reply api/metric/api.go]: Maybe this should be called a `Definition` here, where it is returned as a struct value. In the export package, a pointer is returned allowing export pipelines to use the pointer value as a map key. The export type could be called `Descriptor`. I don't think the user cares too much, maybe there will be some reason a user wants this information, but as an implementor I think this makes sense.

Comment (MrAlias) [reply api/metric/api.go]: > Maybe this should be called a Definition here, where it is returned as a struct value. In the export package, a pointer is returned allowing export pipelines to use the pointer value as a map key. The export type could be called Descriptor. 

This is an interesting idea! I think it is probably better suited for a separate PR, but I like the concept of using a definition to define the instrument and a descriptor in the SDK to describe. That and the customized representation in either sounds like a usability benefit for implementors, SDK writers, and export writers.

File: api/global/internal/meter.go [hunk 18]
```diff
@@ -52,50 +44,41 @@ type meter struct {
...
-	nkind    core.NumberKind
-	opts     []metric.Option
-	meter    *meter
-	callback interface{}
-}
+	instrument
 
-type hasImpl interface {
-	Impl() metric.InstrumentImpl
+	ctor func(metric.Meter) (metric.SynchronousImpl, error)
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply api/global/internal/meter.go]: Ah! TIL :smile:

File: api/metric/common.go [hunk 19]
```diff
@@ -21,45 +21,53 @@ import (
...
+func (h synchronousBoundInstrument) Unbind() {
 	h.boundInstrument.Unbind()
 }
 
-func newCommonMetric(instrument InstrumentImpl, err error) (commonMetric, error) {
+func (a asynchronousInstrument) AsynchronousImpl() AsynchronousImpl {
+	return a.instrument
+}
+
+func checkSynchronous(instrument SynchronousImpl, err error) (synchronousInstrument, error) {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply api/metric/common.go]: SGTM

**Comment by jmacd** [comment]
I want to add that this PR is far from perfect.  One of the decisions that I had trouble making was whether the new `wrappedMeterImpl` in sdkhelpers.go would be used as merely a convenience or whether all SDKs would be required to use it. It seemed like taking a step too far to require SDKs to implement the `MeterImpl` generic constructors (i.e., `NewSynchronousInstrument` and `NewAsynchronousInstrument`) since it could conceivably force them to create generic code they otherwise would not have to / want to. 

This matters for the code in `api/global`, which would be a bit simpler if it could assume that any `Meter` set as the global instance had an underlying `MeterImpl`. This only matters a little bit, however, since the API forces the SDK to provide a `SynchronousImpl` or `AsynchronousImpl` for each instrument constructed. 

As it stands, therefore, the code in `api/global` is not written to assume that the global instance it deals with can be unwrapped. For the global package and any SDK that doesn't want to implement `MeterImpl`, sdkhelpers.go also provides the `Wrap*Instrument` helpers. 

This means that the code in #566 doesn't tell the whole story. The real SDK can acquire uniqueness checking by being wrapped in a new MeterImpl as shown in that PR, but the global package has to expand the access checks into each constructor by hand. So part of me wonders how it would look if: (1) all SDKs were required to implement `MeterImpl`, and (2) the global package was implemented using a wrapped `MeterImpl` instead of a custom implementation as it stands here. This might make uniqueness checking nice, but I'm not sure it will improve readability in the globals package, which is quite complicated.

**Comment by jmacd** [comment]
I added some detail in #566. I now think it won't actually help to assume that all `Meter` implementations are wrapped `MeterImpl`s. I've put some notes into the api/global/internal package where the changes would go, and I think I'm happy with this change (although the PR is too large, sigh).

File: sdk/metric/sdk.go [hunk 20]
```diff
@@ -665,15 +551,16 @@ func (m *SDK) checkpoint(ctx context.Context, descriptor *export.Descriptor, rec
...
 	}
 }
 
-// GetDescriptor returns the descriptor of an instrument, which is not
-// part of the public metric API.
-func (m *SDK) GetDescriptor(inst metric.InstrumentImpl) *export.Descriptor {
-	if ii, ok := inst.(*instrument); ok {
-		return ii.descriptor
+// GetDescriptor returns a pointer to the descriptor of an instrument,
+// which is not part of the public metric API.  This is for testing.  Use
```
Comment (MrAlias) [inline sdk/metric/sdk.go]: If this is for testing and we don't want general users of the package to have access to this shouldn't we not export this?

It seems a bit like we should be dogfooding our design decision here. If we think users won't need this but we need this in other packages it seems like it might be a design problem or a testing scope problem.

File: sdk/metric/batcher/test/test.go [hunk 21]
```diff
@@ -137,14 +139,16 @@ func CounterAgg(desc *export.Descriptor, v int64) export.Aggregator {
...
+
+	if s, ok := rec.Aggregator().(aggregator.Sum); ok {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply sdk/metric/batcher/test/test.go]: Yeah. It felt wrong to be testing for concrete types when the exporters are supposed to test for interfaces, so I changed it.

File: api/global/internal/meter_test.go [hunk 22]
```diff
@@ -300,3 +299,54 @@ func TestErrorInDeferredConstructor(t *testing.T) {
 	c1.Add(ctx, 1, meter.Labels())
 	c2.Add(ctx, 2, meter.Labels())
 }
+
+func TestInterfaceIndirection(t *testing.T) {
+	internal.ResetForTest()
+
+	// Test that Interface() does the proper indirection, i.e.,
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/global/internal/meter_test.go]: Fixed.

File: api/metric/api.go (hunk 17 above)
Comment (jmacd) [reply api/metric/api.go]: Maybe this can be done in #552?

File: internal/metric/mock.go [hunk 23]
```diff
@@ -187,92 +158,25 @@ func (m *Meter) Labels(labels ...core.KeyValue) apimetric.LabelSet {
 	}
 }
 
-func (m *Meter) NewInt64Counter(name string, opts ...apimetric.Option) (apimetric.Int64Counter, error) {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply internal/metric/mock.go]: It was a bit exhausting. Thanks for holding back with #552.

File: sdk/metric/sdk.go (hunk 20 above)
Comment (jmacd) [reply sdk/metric/sdk.go]: This already existed and is only used for testing.  If I moved the test into the same package, we could avoid exposing this. It's used in the stress test and at this moment I'd rather not think about how to eliminate it.

File: sdk/metric/sdk.go (hunk 16 above)
Comment (jmacd) [reply sdk/metric/sdk.go]: Fixed.

File: api/metric/api.go (hunk 17 above)
Comment (jmacd) [reply api/metric/api.go]: 😅

**Comment by jmacd** [comment]
I added a Synchronous->Sync and Asynchronous->Async change. I generally agree, @MrAlias, that a bunch of cleanups and moving things into different files would help. I tried to keep things in place to help the diff be readable. I'd be all for a change that only moves things around for clarity without any functional changes.

File: sdk/metric/sdk.go (hunk 20 above)
Comment (MrAlias) [reply sdk/metric/sdk.go]: Sounds reasonable. +1 for moving forward.

Comment (MrAlias) [reply sdk/metric/sdk.go]: Created https://github.com/open-telemetry/opentelemetry-go/issues/570 to track this.

File: api/metric/api_test.go [hunk 24]
```diff
@@ -117,7 +117,7 @@ func TestOptions(t *testing.T) {
 
 func TestCounter(t *testing.T) {
 	{
-		meter := mock.NewMeter()
+		mock, meter := mockTest.NewMeter()
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/metric/api_test.go]: Done.
